### DISCUSSION TITLE
Different file sizes of the executables (installer vs zip)

### DISCUSSION THREAD

**Discussion by FSys77** [dogecoin/dogecoin#2262] [Help needed]
I just noticed that the file sizes of the executables differ, depending where they are from. For example:

dogecoin-qt.exe:

When taken from the installer (dogecoin-1.14.3-win64-setup-unsigned.exe) the file size is:
33.735.184 bytes

When taken from the zip file (dogecoin-1.14.3-win64.zip) the file size is:
33.735.711 bytes

Same for all the other executables, they all slightly differ.

I checked with older releases, for example 1.10.0, and there the file sizes of the extracted executables are exactly the same in the setup.exe and the .zip version.

Is there something fishy going on recently?

**Comment by rnicoll** [comment]
Interesting point. Honestly, I haven't worked closely on Windows packaging before, most of what I do at my day job is server-side stuff that doesn't ship to customers. I can however tell you how these are built, and how to verify them, and that may help.

We don't want to trust any single machine or developer when building binaries to ship, so instead several developers all build a copy of the release, and sign it, and presuming all of the built copies match, it's good and we ship it. Sometimes they don't match and we have to debug the build environment, which is painful. To ensure the built binaries are consistent, we use a deterministic build tool called [Gitian](https://www.gitian.org/).

You can see the Gitian manifests and their signatures for 1.14.3 under https://github.com/dogecoin/gitian.sigs/tree/master/1.14.3-win (for Windows, there's similar folders for Linux and OS X), and get the GPG keys for the devs from https://github.com/dogecoin/dogecoin/tree/master/contrib/gitian-keys. If you look at the manifest (for example https://github.com/dogecoin/gitian.sigs/blob/master/1.14.3-win/rnicoll/dogecoin-win-1.14-build.assert ) it shows the hashes of the file I built, which should match what you download.

This doesn't explain the file size mismatch, but it should let you verify what you've downloaded is signed off by 4 different people, indepdendently built, to verify it's not been tampered with.

Make sense?

**Reply by FSys77** [reply]
Hello Ross,

thank you very much for your explanation.

(1) Still I would be interested where this file size mismatch comes from. People have started putting serious amounts of money into Dogecoin - and trust is everything.

(2) When I look at the itian.sigs for 1.14.3, those of langerhans and rnicoll are the same. Also michilumin matches with patricklodder. But langerhans/rnicoll are different from michilumin/patricklodder. Why is that?


Regards,
Neil

**Reply by rnicoll** [reply]
On the first point my guess would be https://github.com/dogecoin/dogecoin/blob/master/contrib/gitian-descriptors/gitian-win-signer.yml#L34 where it's taking the unsigned executable and signing it. I think my point though is that anyone is in as good a position as the core maintainers, to verify that (and there's very few of us, so we're trying not to be a bottleneck for everything). If you can find something to verify the packaging structure of a Windows executable and compare the two, that's probably the next step if you want to be absolutely sure.

On the second, it's only the `out_manifest` sections we care about in the final result, because they're what's been built. The `base_manifests` section is the state of the VM we build on, and is useful for diagnosing if the compiled binaries don't match. The VMs are built individually and independently, so I'm more surprised if langerhans & I match exactly, than there being differences. We can't use a single shared VM image because it would mean if that was compromised, we'd be building on a compromised environment. Which is a pity, because it would be a hell of a lot easier if we could, rather than 4 devs having to make 4 different VMs and then debug any differences without being able to log into any other VM.

I do see there's a difference in the `in_manifest` section, that's more confusing, it looks like I had the wrong build script version. That's my fault, but again as we all 4 ended up with the same result, as long as you trust the other 3, the weirdness in my input script shouldn't be a major concern.

Aside; we always need more people doing the Gitian build & sign, if you want to be part of the verification process for future releases?

**Reply by rnicoll** [reply]
Have a read over https://github.com/dogecoin/dogecoin/blob/master/doc/gitian-building.md for what that would entail. It's a pain getting the VM set up, but then in theory straight forward.

**Reply by FSys77** [reply]
> Aside; we always need more people doing the Gitian build & sign, if you want to be part of the verification process for future releases?

Yes, why not :-D

I have been a C developer in a previous life, so having a brief look over the source code should be feasible. Setting up VMs is also something I am familiar with.

Curiosity: How is the build & release process for this project? And stuff like compliance, security audits, etc... I mean we talk about billions of dollars here.

**Reply by rnicoll** [reply]
So it's all open source volunteer basis, so as with any other critical open source project, we mostly rely on many eyes. There's four of us who review everything that goes in. We do get informal security reviews from researchers (1.14.3 for example fixed a security issue that was raised an external).

Release process then is basically one of us creates a tag, the others check the history on that tag, we then independently build that tagged version, push the signatures to the GitHub repository, and if they all match we push to GitHub releases, and then refresh the website. We're trying to find ways to scale, but the very high bar for volunteers to be able to review, because it's a multi-billion dollar currency, makes that exception hard. Keep in mind we have no revenue stream (at least currently), so full time employees aren't a thing. Suggestions welcomed, of course.

**Reply by patricklodder** [reply]
> (2) When I look at the [g]itian.sigs for 1.14.3, those of langerhans and rnicoll are the same. Also michilumin matches with patricklodder. But langerhans/rnicoll are different from michilumin/patricklodder. Why is that?

What differences did you spot exactly? Because to me, the win outputs look to be the same for all, the input manifests are the same for everyone except Ross (different descriptor version?) and the environments always differ (but that's ok) except between Michi and myself because we used the exact same procedure on the same day ([this one](https://gist.github.com/patricklodder/fcffff8af0df78bdfb2957581c7368e5)) because we both needed a new gitian builder install.

**Reply by FSys77** [reply]
> Aside; we always need more people doing the Gitian build & sign, if you want to be part of the verification process for future releases?

OK. I've set up the Gitian VM. How can I get involved in the next release?

**Reply by rnicoll** [reply]
You can always retrospectively build and validate 1.14.3 as practice. First step would be to build 1.14.3 and then confirm you can sign, as that's trickiest. We'd then get you to upload you GPG public key, and finally your signatures. As we can't verify the identity on the public key, it'll have limited trust, but everything helps.

**Reply by rnicoll** [reply]
Try a command like:

```
./bin/gbuild --memory 3000 --commit dogecoin=1.14.3 ../dogecoin/contrib/gitian-descriptors/gitian-osx.yml -j8
```

Replacing the descriptor with the platform you want to target, to get started building. `-j8` should match your number of CPU cores.
